File: 21. Merge Two Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode* head = NULL;
        ListNode* node = NULL;
        int val;
        while( l1 != NULL || l2 != NULL ) {
            
           if( l1 && l2 && l1->val <= l2->val ) { 
                val = l1->val;
                l1 = l1->next;
           } else if( l1 && l2 && l1->val > l2->val ) { 
                val = l2->val; 
                l2 = l2->next;
          } else if( l1 ) { 
                val = l1->val;
                l1 = l1->next;
          } else { 
                val = l2->val;
                l2 = l2->next;
          }
          ListNode* curNode = new ListNode(val);
          if( node ) { 
                node->next = curNode;
                node = node->next;
          } else { 
                node = curNode;
                head = node;
          }
        }
    return head;
    }
};

File: python/Lakshay/21. Merge Two Sorted Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        ans = None
        prv = None
        while l1 and l2:
            curNode = None
            if l1.val <= l2.val:
                curNode = ListNode(l1.val)
                l1 = l1.next
            else:
                curNode = ListNode(l2.val)
                l2 = l2.next
            
            if prv is None:
                ans = curNode
            else:
                prv.next = curNode
            
            prv = curNode
            
        while l1:
            curNode = ListNode(l1.val)
            l1 = l1.next
            if prv is None:
                ans = curNode
            else:
                prv.next = curNode
            prv = curNode
        
        while l2:
            curNode = ListNode(l2.val)
            l2 = l2.next
            if prv is None:
                ans = curNode
            else:
                prv.next = curNode
            prv = curNode
            
        return ans
                
                
        
